### PR TITLE
[Toolchain] Add TypeScript support 

### PR DESCRIPTION
**Pull request by damassi**
This PR adds TypeScript into the mix via `ts-loader`. Additional additions are Prettier. Very little has changed, though there is a slight difference between how errors are handled in normal JS code vs TypeScript code, related to general incompatibility between [friendly-errors-webpack-plugin](https://github.com/geowarin/friendly-errors-webpack-plugin) and TSC, though I suspect this can be fixed at a later date. That said, codeframe errors in the terminal provide most of the functionality so that errors still remain useful and pretty:

screen shot 2018-02-09 at 4 22 43 pm
 
Also added `cache-loader` which has halved our boot times and brought incremental compilation down to around ~1-2 sec, regardless of file-type.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,12 +1,18 @@
-global.Promise = require('bluebird')
+require('babel-core/register', {
+  extensions: ['.js', '.jsx', '.ts', '.tsx'],
+})
+
+require('ts-node').register()
```
Comment (damassi) [inline index.js]: This worked well for SSR stuff

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a PR summary</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: desktop/apps/article/queries/articleBody.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-export default `
+export const articleBody = `
```
Comment (orta) [inline desktop/apps/article/queries/articleBody.js]: nice move

File: package.json [hunk 3]
```diff
@@ -12,11 +12,26 @@
...
+    "prettier-write": "yarn run prettier --write",
     "start": "sh scripts/start.sh",
     "test": "sh scripts/test.sh",
     "webpack": "webpack"
   },
+  "prettier": {
+    "arrowParens": "always",
+    "bracketSpacing": true,
+    "semi": false,
+    "singleQuote": true,
```
Comment (damassi) [inline package.json]: Sorry y'all 😄 we type less JSON than JS and our wrist muscles will thank us later

**Comment by damassi** [comment]
The ease and stability of this route makes me want to update Prediction similarly.

File: tslint.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": [
+    "tslint-config-standard",
+    "tslint-config-prettier",
+    "tslint-react"
+  ],
+  "rules": {
+    "space-before-function-paren": false
+  }
```
Comment (orta) [inline tslint.json]: https://palantir.github.io/tslint/rules/no-default-export/

Especially as there is no TS right now

File: desktop/apps/article/queries/relatedArticles.js [hunk 5]
```diff
@@ -1,4 +1,4 @@
-export default `
+export const relatedArticles = `
   relatedArticlesPanel {
     slug
     thumbnail_title
```
Comment (damassi) [inline desktop/apps/article/queries/relatedArticles.js]: @kanaabe - I noticed that this was throwing strange errors; it looks like the template literal needs a name

Comment (alloy) [reply desktop/apps/article/queries/relatedArticles.js]: Just out of curiosity, what tooling wants this to have a name?

Comment (damassi) [reply desktop/apps/article/queries/relatedArticles.js]: Looking now, it seemed like 
```
export default ` ... `
``` 

was a transient issue -- trying now, everything is 👍

File: index.js [hunk 6]
```diff
@@ -1,12 +1,15 @@
-global.Promise = require('bluebird')
-require('coffee-script/register')
 require('babel-core/register')
+require('ts-node').register()
+require('coffee-script/register')
```
Comment (alloy) [inline index.js]: Am I correct in understanding that we’re _only_ AOT compiling client-side code? Do you have any insights into how much memory all of these are using at runtime once they’ve lazily loaded all the files?

Comment (alloy) [reply index.js]: To clarify, I would really like us to AOT server-side code too.

File: desktop/apps/article/queries/relatedArticles.js (hunk 5 above)
Comment (alloy) [reply desktop/apps/article/queries/relatedArticles.js]: Okeydokey. I’m not actually in favour of anonymous exports btw ;)

File: index.js (hunk 6 above)
Comment (damassi) [reply index.js]: Yup, only client-side code, and will need to measure server-side memory usage. There is a doc that Cavvia and Joey and I have been working on [located here](https://docs.google.com/document/u/1/d/1818OAyc76N1jasRBuRNJFazNMXHHJId_rttF76RfG-k/edit) that we can add this line item to and investigate.

**Comment by alloy** [comment]
This all looks good to me, great job @damassi 👏 💯 

I have just two questions:
* What is the (memory) impact of using JIT compilation in production?
* Is adding Sentry source-maps uploading on the roadmap for this project? https://docs.sentry.io/clients/node/sourcemaps/

**Comment by damassi** [comment]
@alloy - I'll go ahead and get those stats now, and will add in the Sentry Webpack loader. Great idea.

**Comment by alloy** [comment]
Ace, thanks!

FYI, I think I linked to the Sentry docs on source-maps for server-side code, be sure to look at the right ones instead ;)

**Comment by damassi** [comment]
@alloy - Looking [here](https://dashboard.heroku.com/apps/force-production/metrics/web?starting=168-hours-ago) it seems like our average memory usage in force is around 370mb, which seems to be nice and steady and relatively low. I can instrument things so that we can get precise metrics around the registration hooks, but I think then end goal should be AOT compilation, in another PR, which will mostly affect performance. I can handle that after this sprint's other tasks are complete.

**Comment by alloy** [comment]
Sounds good to me 👍

**Pull request merged**